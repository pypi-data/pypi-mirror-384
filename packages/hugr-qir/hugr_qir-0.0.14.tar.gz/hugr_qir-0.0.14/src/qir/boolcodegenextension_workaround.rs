use anyhow::anyhow;
use hugr_core::ops::{ExtensionOp, Value};
use hugr_core::types::{SumType, TypeName};
use hugr_core::{HugrView, Node};
use hugr_llvm::CodegenExtension;
use hugr_llvm::emit::{EmitFuncContext, EmitOpArgs, emit_value};
use hugr_llvm::inkwell::IntPredicate;
use hugr_llvm::inkwell::types::IntType;
use hugr_llvm::sum::LLVMSumValue;
use hugr_llvm::types::TypingSession;
use tket::extension::bool::{BOOL_EXTENSION_ID, BoolOp, ConstBool};

// This is copied from https://github.com/CQCL/guppylang/blob/main/execute_llvm/src/bool.rs
// And is a workaround until the below issues are resolved
// https://github.com/CQCL/tket2/issues/909
// https://github.com/CQCL/tket2/issues/910

#[allow(dead_code)]
const BOOL_TYPE_ID: TypeName = TypeName::new_inline("bool");

#[allow(dead_code)]
fn llvm_bool_type<'c>(ts: &TypingSession<'c, '_>) -> IntType<'c> {
    ts.iw_context().bool_type()
}
#[allow(dead_code)]
pub struct BoolCodegenExtension;

impl BoolCodegenExtension {
    #[allow(dead_code)]
    fn emit_bool_op<'c, H: HugrView<Node = Node>>(
        &self,
        context: &mut EmitFuncContext<'c, '_, H>,
        args: EmitOpArgs<'c, '_, ExtensionOp, H>,
        op: BoolOp,
    ) -> anyhow::Result<()> {
        match op {
            BoolOp::read => {
                let [inp] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("BoolOp::read expects one argument"))?;
                let res = inp.into_int_value();
                let true_val = emit_value(context, &Value::true_val())?;
                let false_val = emit_value(context, &Value::false_val())?;
                let res = context
                    .builder()
                    .build_select(res, true_val, false_val, "")?;
                args.outputs.finish(context.builder(), vec![res])
            }
            BoolOp::make_opaque => {
                let [inp] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("BoolOp::make_opaque expects one argument"))?;
                let bool_ty = context.llvm_sum_type(SumType::new_unary(2))?;
                let bool_val = LLVMSumValue::try_new(inp, bool_ty)?;
                let res = bool_val.build_get_tag(context.builder())?;
                args.outputs.finish(context.builder(), vec![res.into()])
            }
            BoolOp::not => {
                let [inp] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("BoolOp::not expects one argument"))?;
                let res = inp.into_int_value();
                let res = context.builder().build_not(res, "")?;
                args.outputs.finish(context.builder(), vec![res.into()])
            }
            binary_op => {
                let [inp1, inp2] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("BoolOp::{:?} expects two arguments", binary_op))?;
                let inp1_val = inp1.into_int_value();
                let inp2_val = inp2.into_int_value();
                let res = match binary_op {
                    BoolOp::and => context.builder().build_and(inp1_val, inp2_val, "")?,
                    BoolOp::or => context.builder().build_or(inp1_val, inp2_val, "")?,
                    BoolOp::xor => context.builder().build_xor(inp1_val, inp2_val, "")?,
                    BoolOp::eq => context.builder().build_int_compare(
                        IntPredicate::EQ,
                        inp1_val,
                        inp2_val,
                        "",
                    )?,
                    _ => return Err(anyhow!("Unsupported binary bool operation")),
                };
                args.outputs.finish(context.builder(), vec![res.into()])
            }
        }
    }
}

impl CodegenExtension for BoolCodegenExtension {
    fn add_extension<'a, H: hugr::HugrView<Node = hugr::Node> + 'a>(
        self,
        builder: hugr::llvm::CodegenExtsBuilder<'a, H>,
    ) -> hugr::llvm::CodegenExtsBuilder<'a, H>
    where
        Self: 'a,
    {
        builder
            .custom_type((BOOL_EXTENSION_ID, BOOL_TYPE_ID), |ts, _| {
                Ok(llvm_bool_type(&ts).into())
            })
            .custom_const::<ConstBool>(|context, val| {
                let bool_ty = llvm_bool_type(&context.typing_session());
                Ok(bool_ty.const_int(val.value().into(), false).into())
            })
            .simple_extension_op(move |context, args, op| self.emit_bool_op(context, args, op))
    }
}
