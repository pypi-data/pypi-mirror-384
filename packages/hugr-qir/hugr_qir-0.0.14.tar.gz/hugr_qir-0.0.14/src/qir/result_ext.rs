use anyhow::{Result, anyhow, bail};
use hugr::{
    HugrView, Node,
    extension::{prelude::ConstString, simple_op::MakeExtensionOp as _},
    ops::ExtensionOp,
};
use hugr_llvm::{
    emit::{EmitFuncContext, EmitOpArgs, emit_value},
    inkwell::types::BasicType as _,
    sum::LLVMSumValue,
    types::HugrSumType,
};
use tket_qsystem::extension::result::{ResultOp, ResultOpDef};

use super::QirCodegenExtension;
impl QirCodegenExtension {
    pub fn emit_result_op<'c, H: HugrView<Node = Node>>(
        &self,
        context: &mut EmitFuncContext<'c, '_, H>,
        args: EmitOpArgs<'c, '_, ExtensionOp, H>,
        op: ResultOpDef,
    ) -> Result<()> {
        let result_op = ResultOp::from_extension_op(&args.node())?;
        let tag_str = result_op.tag;
        if tag_str.is_empty() {
            bail!("Empty result tag received")
        }

        let i8_ptr_ty = context
            .iw_context()
            .i8_type()
            .ptr_type(Default::default())
            .as_basic_type_enum();
        let tag_ptr = {
            let x = emit_value(context, &ConstString::new(tag_str).into())?;
            if x.get_type() == i8_ptr_ty {
                x
            } else {
                context.builder().build_bit_cast(x, i8_ptr_ty, "tag_ptr")?
            }
        };

        match op {
            ResultOpDef::Bool => {
                let [val] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("result_bool expects one input"))?;
                let bool_type = context.llvm_sum_type(HugrSumType::new_unary(2))?;
                let val = LLVMSumValue::try_new(val, bool_type)
                    .map_err(|_| anyhow!("bool_type expects a value"))?
                    .build_get_tag(context.builder())?;
                let i1_ty = context.iw_context().bool_type();
                let trunc_val = context.builder().build_int_truncate(val, i1_ty, "")?;
                let print_fn_ty = context
                    .iw_context()
                    .void_type()
                    .fn_type(&[i1_ty.into(), i8_ptr_ty.into()], false);
                let print_fn =
                    context.get_extern_func("__quantum__rt__bool_record_output", print_fn_ty)?;
                context.builder().build_call(
                    print_fn,
                    &[trunc_val.into(), tag_ptr.into()],
                    "print_bool",
                )?;
                args.outputs.finish(context.builder(), [])
            }
            ResultOpDef::Int | ResultOpDef::UInt => {
                let [mut val] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("result_bool expects one input"))?;
                let i64_ty = context.iw_context().i64_type();
                if val.get_type() != i64_ty.into() {
                    val = if op == ResultOpDef::Int {
                        context
                            .builder()
                            .build_int_s_extend(val.into_int_value(), i64_ty, "")
                    } else {
                        context
                            .builder()
                            .build_int_z_extend(val.into_int_value(), i64_ty, "")
                    }?
                    .into();
                }
                let print_fn_ty = context
                    .iw_context()
                    .void_type()
                    .fn_type(&[i64_ty.into(), i8_ptr_ty.into()], false);
                let print_fn =
                    context.get_extern_func("__quantum__rt__int_record_output", print_fn_ty)?;
                context.builder().build_call(
                    print_fn,
                    &[val.into(), tag_ptr.into()],
                    "print_bool",
                )?;
                args.outputs.finish(context.builder(), [])
            }
            ResultOpDef::F64 => {
                let [val] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("result_bool expects one input"))?;
                let f64_ty = context.iw_context().f64_type();
                let print_fn_ty = context
                    .iw_context()
                    .void_type()
                    .fn_type(&[f64_ty.into(), i8_ptr_ty.into()], false);
                let print_fn =
                    context.get_extern_func("__quantum__rt__double_record_output", print_fn_ty)?;
                context.builder().build_call(
                    print_fn,
                    &[val.into(), tag_ptr.into()],
                    "print_bool",
                )?;
                args.outputs.finish(context.builder(), [])
            }
            // these ops are not supported yet
            ResultOpDef::ArrBool
            | ResultOpDef::ArrInt
            | ResultOpDef::ArrUInt
            | ResultOpDef::ArrF64
            | _ => bail!("Unknown op: {op:?}"),
        }
    }
}

#[cfg(test)]
mod test {
    use hugr::ops::OpType;
    use hugr_llvm::{
        check_emission,
        test::{TestContext, llvm_ctx},
    };
    use rstest::rstest;

    use tket_qsystem::extension::result::ResultOpDef;

    use crate::qir::{QirCodegenExtension, QirPreludeCodegen};
    use crate::test::single_op_hugr;

    #[rstest::fixture]
    fn ctx(mut llvm_ctx: TestContext) -> TestContext {
        llvm_ctx.add_extensions(|builder| {
            builder
                .add_extension(QirCodegenExtension)
                .add_prelude_extensions(QirPreludeCodegen)
                .add_default_int_extensions()
                .add_float_extensions()
        });
        llvm_ctx
    }

    #[rstest]
    #[case(ResultOpDef::F64.instantiate(&["foo_f64".into()]).unwrap())]
    #[case(ResultOpDef::UInt.instantiate(&["foo_uint".into(), 3.into()]).unwrap())]
    #[case(ResultOpDef::Int.instantiate(&["foo_int".into(), 4.into()]).unwrap())]
    #[case(ResultOpDef::Bool.instantiate(&["bool_int".into()]).unwrap())]
    fn emit(ctx: TestContext, #[case] op: impl Into<OpType>) {
        let op = op.into();
        let mut insta = insta::Settings::clone_current();
        insta.set_snapshot_suffix(format!("{}_{}", insta.snapshot_suffix().unwrap_or(""), op));
        insta.bind(|| {
            let hugr = single_op_hugr(op);
            check_emission!(hugr, ctx);
        })
    }
}
