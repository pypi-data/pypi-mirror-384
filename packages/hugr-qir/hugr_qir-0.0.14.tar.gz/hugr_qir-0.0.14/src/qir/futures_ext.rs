use anyhow::{Result, bail, ensure};
use hugr::{
    HugrView, Node,
    extension::prelude::bool_t,
    ops::{ExtensionOp, Value},
    types::CustomType,
};
use hugr_llvm::{
    emit::{EmitFuncContext, EmitOpArgs, emit_value},
    inkwell::types::BasicTypeEnum,
    types::TypingSession,
};
use tket_qsystem::extension::futures::{EXTENSION_ID, FUTURE_TYPE_NAME, FutureOpDef};

use super::QirCodegenExtension;

impl QirCodegenExtension {
    /// We represent a hugr `tket.futures.future<bool>` as an i1.
    pub fn convert_future_type<'c>(
        &self,
        session: TypingSession<'c, '_>,
        custom_type: &CustomType,
    ) -> Result<BasicTypeEnum<'c>> {
        ensure!(
            custom_type.extension() == &EXTENSION_ID
                && custom_type.name() == FUTURE_TYPE_NAME.as_str()
                && custom_type.args() == [bool_t().into()]
        );
        Ok(session.iw_context().bool_type().into())
    }

    pub fn emit_futures_op<'c, H: HugrView<Node = Node>>(
        &self,
        context: &mut EmitFuncContext<'c, '_, H>,
        args: EmitOpArgs<'c, '_, ExtensionOp, H>,
        op: FutureOpDef,
    ) -> Result<()> {
        match op {
            FutureOpDef::Read => {
                let true_val = emit_value(context, &Value::true_val())?;
                let false_val = emit_value(context, &Value::false_val())?;

                let bool_r = context.builder().build_select(
                    args.inputs[0].into_int_value(),
                    true_val,
                    false_val,
                    "",
                )?;
                args.outputs.finish(context.builder(), [bool_r])
            }
            FutureOpDef::Dup => {
                let input = args.inputs[0];
                args.outputs.finish(context.builder(), [input, input])
            }
            FutureOpDef::Free => args.outputs.finish(context.builder(), []),
            _ => bail!("Unknown op: {op:?}"),
        }
    }
}

#[cfg(test)]
mod test {
    use hugr::extension::simple_op::HasConcrete as _;
    use hugr::{extension::prelude::bool_t, ops::OpType};
    use hugr_llvm::{
        check_emission,
        test::{TestContext, llvm_ctx},
    };
    use rstest::rstest;

    use tket_qsystem::extension::futures::FutureOpDef;

    use crate::qir::{QirCodegenExtension, QirPreludeCodegen};
    use crate::test::single_op_hugr;

    #[rstest::fixture]
    fn ctx(mut llvm_ctx: TestContext) -> TestContext {
        llvm_ctx.add_extensions(|builder| {
            builder
                .add_extension(QirCodegenExtension)
                .add_prelude_extensions(QirPreludeCodegen)
        });
        llvm_ctx
    }

    #[rstest]
    #[case(FutureOpDef::Read.instantiate(&[bool_t().into()]).unwrap())]
    #[case(FutureOpDef::Dup.instantiate(&[bool_t().into()]).unwrap())]
    #[case(FutureOpDef::Free.instantiate(&[bool_t().into()]).unwrap())]
    fn emit(ctx: TestContext, #[case] op: impl Into<OpType>) {
        let op = op.into();
        let mut insta = insta::Settings::clone_current();
        insta.set_snapshot_suffix(format!("{}_{}", insta.snapshot_suffix().unwrap_or(""), op));
        insta.bind(|| {
            let hugr = single_op_hugr(op);
            check_emission!(hugr, ctx);
        })
    }
}
