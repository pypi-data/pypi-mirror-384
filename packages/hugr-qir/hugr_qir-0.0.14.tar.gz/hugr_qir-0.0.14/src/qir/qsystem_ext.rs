use crate::qir::{
    emit_qis_gate, emit_qis_gate_finish, emit_qis_measure_to_result, emit_qis_qalloc,
    emit_qis_qfree, emit_qis_read_result,
};
use anyhow::Result;
use hugr::{
    HugrView, Node,
    extension::prelude::{option_type, qb_t},
    ops::ExtensionOp,
};
use hugr_llvm::{
    emit::{EmitFuncContext, EmitOpArgs},
    sum::LLVMSumValue,
    types::HugrSumType,
};
use tket_qsystem::extension::qsystem::QSystemOp;

use super::QirCodegenExtension;

impl QirCodegenExtension {
    pub fn emit_qsystem_op<'c, H: HugrView<Node = Node>>(
        &self,
        context: &mut EmitFuncContext<'c, '_, H>,
        args: EmitOpArgs<'c, '_, ExtensionOp, H>,
        op: QSystemOp,
    ) -> Result<()> {
        use QSystemOp::*;
        match op {
            Measure => {
                let qb = args.inputs[0];
                // i.e. Result*
                let result = emit_qis_measure_to_result(context, qb)?;

                let result_bool = emit_qis_read_result(context, result)?;
                args.outputs.finish(context.builder(), [result_bool])
            }
            LazyMeasure => {
                let qb = args.inputs[0];
                // i.e. Result*
                let result = emit_qis_measure_to_result(context, qb)?;

                let result_bool = emit_qis_read_result(context, result)?;
                let result_sum = LLVMSumValue::try_new(
                    result_bool,
                    context.llvm_sum_type(HugrSumType::new_unary(2))?,
                )?;

                let result_i32 = result_sum.build_get_tag(context.builder())?;
                let i1 = context.iw_context().bool_type();
                let result_i1 = context.builder().build_int_truncate(result_i32, i1, "")?;
                // futures are i1s, so this is fine
                args.outputs.finish(context.builder(), [result_i1.into()])
            }
            MeasureReset => {
                let qb = args.inputs[0];
                // i.e. Result*
                let result = emit_qis_measure_to_result(context, qb)?;
                let _ = emit_qis_gate(context, "__quantum__qis__reset__body", [], [qb])?;
                let result_bool = emit_qis_read_result(context, result)?;
                args.outputs.finish(context.builder(), [qb, result_bool])
            }
            LazyMeasureReset => {
                let qb = args.inputs[0];
                // i.e. Result*
                let result = emit_qis_measure_to_result(context, qb)?;
                let _ = emit_qis_gate(context, "__quantum__qis__reset__body", [], [qb])?;
                let result_bool = emit_qis_read_result(context, result)?;
                let result_sum = LLVMSumValue::try_new(
                    result_bool,
                    context.llvm_sum_type(HugrSumType::new_unary(2))?,
                )?;
                let result_i32 = result_sum.build_get_tag(context.builder())?;
                let i1 = context.iw_context().bool_type();
                let result_i1 = context.builder().build_int_truncate(result_i32, i1, "")?;
                args.outputs
                    .finish(context.builder(), [qb, result_i1.into()])
            }
            Rz => emit_qis_gate_finish(
                context,
                "__quantum__qis__rz__body",
                &args.inputs[1..2],
                &args.inputs[0..1],
                args.outputs,
            ),
            PhasedX => emit_qis_gate_finish(
                context,
                "__quantum__qis__phasedx__body",
                &args.inputs[1..3],
                &args.inputs[0..1],
                args.outputs,
            ),
            ZZPhase => emit_qis_gate_finish(
                context,
                "__quantum__qis__rzz__body",
                &args.inputs[2..3],
                &args.inputs[0..2],
                args.outputs,
            ),
            TryQAlloc => {
                let qb = emit_qis_qalloc(context)?;
                let option_ty = context.llvm_sum_type(option_type(qb_t()))?;
                let qb = option_ty.build_tag(context.builder(), 1, vec![qb])?.into();
                args.outputs.finish(context.builder(), [qb])
            }
            QFree => emit_qis_qfree(context, args.inputs[0]),
            Reset => emit_qis_gate_finish(
                context,
                "__quantum__qis__reset__body",
                [],
                &args.inputs,
                args.outputs,
            ),
            _ => anyhow::bail!("Unknown op: {op:?}"),
        }
    }
}

#[cfg(test)]
mod test {

    use hugr::ops::OpType;
    use hugr_llvm::{
        check_emission,
        test::{TestContext, llvm_ctx},
    };
    use rstest::rstest;

    use tket_qsystem::extension::qsystem::QSystemOp;

    use crate::qir::{QirCodegenExtension, QirPreludeCodegen};
    use crate::test::single_op_hugr;

    #[rstest::fixture]
    fn ctx(mut llvm_ctx: TestContext) -> TestContext {
        llvm_ctx.add_extensions(|builder| {
            builder
                .add_extension(QirCodegenExtension)
                .add_prelude_extensions(QirPreludeCodegen)
                .add_default_int_extensions()
                .add_float_extensions()
        });
        llvm_ctx
    }

    #[rstest]
    #[case(QSystemOp::MeasureReset)]
    #[case(QSystemOp::Reset)]
    #[case(QSystemOp::QFree)]
    #[case(QSystemOp::TryQAlloc)]
    #[case(QSystemOp::ZZPhase)]
    #[case(QSystemOp::PhasedX)]
    #[case(QSystemOp::Rz)]
    #[case(QSystemOp::LazyMeasure)]
    #[case(QSystemOp::LazyMeasureReset)]
    #[case(QSystemOp::Measure)]
    fn emit(ctx: TestContext, #[case] op: impl Into<OpType>) {
        let op = op.into();
        let mut insta = insta::Settings::clone_current();
        insta.set_snapshot_suffix(format!("{}_{}", insta.snapshot_suffix().unwrap_or(""), op));
        insta.bind(|| {
            use tket_qsystem::QSystemPass;

            let mut hugr = single_op_hugr(op);
            QSystemPass::default().run(&mut hugr).unwrap();
            check_emission!(hugr, ctx);
        })
    }
}
