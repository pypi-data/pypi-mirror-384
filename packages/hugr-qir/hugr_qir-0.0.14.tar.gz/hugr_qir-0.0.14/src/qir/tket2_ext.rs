use crate::qir::{
    emit_qis_gate_finish, emit_qis_measure_to_result, emit_qis_qalloc, emit_qis_qfree,
    emit_qis_read_result,
};
use anyhow::{Result, bail};
use hugr::{HugrView, Node, ops::ExtensionOp};
use hugr_llvm::emit::{EmitFuncContext, EmitOpArgs};

use super::QirCodegenExtension;

impl QirCodegenExtension {
    #[allow(non_snake_case)]
    pub fn emit_tket_op<'c, H: HugrView<Node = Node>>(
        &self,
        context: &mut EmitFuncContext<'c, '_, H>,
        args: EmitOpArgs<'c, '_, ExtensionOp, H>,
        op: tket::TketOp,
    ) -> Result<()> {
        use tket::TketOp::*;
        match op {
            H => emit_qis_gate_finish(
                context,
                "__quantum__qis__h__body",
                [],
                args.inputs,
                args.outputs,
            ),
            CX => emit_qis_gate_finish(
                context,
                "__quantum__qis__cx__body",
                [],
                args.inputs,
                args.outputs,
            ),
            CY => emit_qis_gate_finish(
                context,
                "__quantum__qis__cy__body",
                [],
                args.inputs,
                args.outputs,
            ),
            CZ => emit_qis_gate_finish(
                context,
                "__quantum__qis__cz__body",
                [],
                args.inputs,
                args.outputs,
            ),
            T => emit_qis_gate_finish(
                context,
                "__quantum__qis__t__body",
                [],
                args.inputs,
                args.outputs,
            ),
            Tdg => emit_qis_gate_finish(
                context,
                "__quantum__qis__t__adj",
                [],
                args.inputs,
                args.outputs,
            ),
            S => emit_qis_gate_finish(
                context,
                "__quantum__qis__s__body",
                [],
                args.inputs,
                args.outputs,
            ),
            Sdg => emit_qis_gate_finish(
                context,
                "__quantum__qis__s__adj",
                [],
                args.inputs,
                args.outputs,
            ),
            X => emit_qis_gate_finish(
                context,
                "__quantum__qis__x__body",
                [],
                args.inputs,
                args.outputs,
            ),
            Y => emit_qis_gate_finish(
                context,
                "__quantum__qis__y__body",
                [],
                args.inputs,
                args.outputs,
            ),
            Z => emit_qis_gate_finish(
                context,
                "__quantum__qis__z__body",
                [],
                args.inputs,
                args.outputs,
            ),
            Rx => emit_qis_gate_finish(
                context,
                "__quantum__qis__rx__body",
                &args.inputs[1..2],
                &args.inputs[0..1],
                args.outputs,
            ),
            Ry => emit_qis_gate_finish(
                context,
                "__quantum__qis__ry__body",
                &args.inputs[1..2],
                &args.inputs[0..1],
                args.outputs,
            ),
            Rz => emit_qis_gate_finish(
                context,
                "__quantum__qis__rz__body",
                &args.inputs[1..2],
                &args.inputs[0..1],
                args.outputs,
            ),
            Reset => emit_qis_gate_finish(
                context,
                "__quantum__qis__reset__body",
                [],
                &args.inputs,
                args.outputs,
            ),
            Measure => {
                let qb = args.inputs[0];
                // i.e. Result*
                let result = emit_qis_measure_to_result(context, qb)?;

                let result_i1 = emit_qis_read_result(context, result)?;
                args.outputs.finish(context.builder(), [qb, result_i1])
            }
            MeasureFree => {
                let qb = args.inputs[0];
                // i.e. Result*
                let result = emit_qis_measure_to_result(context, qb)?;
                emit_qis_qfree(context, qb)?;
                let result_i1 = emit_qis_read_result(context, result)?;
                args.outputs.finish(context.builder(), [result_i1])
            }
            QAlloc => {
                let qb = emit_qis_qalloc(context)?;
                args.outputs.finish(context.builder(), [qb])
            }
            QFree => emit_qis_qfree(context, args.inputs[0]),
            _ => bail!("Unknown op: {op:?}"),
        }
    }
}

#[cfg(test)]
mod test {
    use hugr::ops::OpType;
    use hugr_llvm::{
        check_emission,
        test::{TestContext, llvm_ctx},
    };
    use rstest::rstest;
    use tket::TketOp;

    use crate::qir::boolcodegenextension_workaround::BoolCodegenExtension;
    use crate::test::single_op_hugr;
    use crate::{
        qir::{QirCodegenExtension, QirPreludeCodegen},
        rotation::RotationCodegenExtension,
    };

    #[rstest::fixture]
    fn ctx(mut llvm_ctx: TestContext) -> TestContext {
        llvm_ctx.add_extensions(|builder| {
            builder
                .add_extension(QirCodegenExtension)
                .add_prelude_extensions(QirPreludeCodegen)
                .add_extension(RotationCodegenExtension::new(QirPreludeCodegen))
                .add_extension(BoolCodegenExtension)
        });
        llvm_ctx
    }

    #[rstest]
    #[case(TketOp::QFree)]
    #[case(TketOp::QAlloc)]
    #[case(TketOp::MeasureFree)]
    #[case(TketOp::Measure)]
    #[case(TketOp::Reset)]
    #[case(TketOp::Rz)]
    #[case(TketOp::Ry)]
    #[case(TketOp::Rx)]
    #[case(TketOp::Z)]
    #[case(TketOp::Y)]
    #[case(TketOp::X)]
    #[case(TketOp::Sdg)]
    #[case(TketOp::S)]
    #[case(TketOp::Tdg)]
    #[case(TketOp::T)]
    #[case(TketOp::CX)]
    #[case(TketOp::CY)]
    #[case(TketOp::CZ)]
    fn emit(ctx: TestContext, #[case] op: impl Into<OpType>) {
        let op = op.into();
        let mut insta = insta::Settings::clone_current();
        insta.set_snapshot_suffix(format!("{}_{}", insta.snapshot_suffix().unwrap_or(""), op));
        insta.bind(|| {
            let hugr = single_op_hugr(op);
            check_emission!(hugr, ctx);
        })
    }
}
