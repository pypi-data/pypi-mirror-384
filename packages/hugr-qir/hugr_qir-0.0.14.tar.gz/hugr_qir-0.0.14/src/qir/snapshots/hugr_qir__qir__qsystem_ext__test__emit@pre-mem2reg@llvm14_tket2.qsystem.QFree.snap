---
source: src/qir/qsystem_ext.rs
expression: mod_str
---
; ModuleID = 'test_context'
source_filename = "test_context"

%Qubit = type opaque

define void @_hl.main.1(%Qubit* %0) {
alloca_block:
  %"2_0" = alloca %Qubit*, align 8
  br label %entry_block

entry_block:                                      ; preds = %alloca_block
  store %Qubit* %0, %Qubit** %"2_0", align 8
  %"2_01" = load %Qubit*, %Qubit** %"2_0", align 8
  call void @__quantum__rt__qubit_release(%Qubit* %"2_01")
  ret void
}

declare void @__quantum__rt__qubit_release(%Qubit*)
